### PR TITLE
Static code analysis with Php Inspections (EA Extended)

### PR DESCRIPTION
**Pull request by kalessil**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | --
| License       | MIT
| Doc PR        | --

- Unnecessary proxy methods and methods duplicating  parents ones removed;
- Type casings instead of PHP4 functions;
- Usage of copy instead of file combining file get/put contents;
- One time use variables cleaned up;
- Superfluous functions calls in loop termination conditions fixed;

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/DataCollector/Util/ValueExporter.php [hunk 1]
```diff
@@ -33,6 +33,7 @@ public function exportValue($value, $depth = 1, $deep = false)
 
         if (is_object($value)) {
             if ($value instanceof \DateTime || $value instanceof \DateTimeInterface) {
+                // note: \DateTime::ISO8601 is not ISO8601 compatible; \DateTime::ATOM IS ISO8601 compatible
                 return sprintf('Object(%s) - %s', get_class($value), $value->format(\DateTime::ISO8601));
```
Comment (kalessil) [inline src/Symfony/Component/HttpKernel/DataCollector/Util/ValueExporter.php]: Perhaps someone fro the core team can give a go fixing this. I'm simply not sure if the change would introduce a BC break...

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SimpleFormFactory.php [hunk 2]
```diff
@@ -77,17 +77,4 @@ protected function createListener($container, $id, $config, $userProvider)
 
         return $listenerId;
     }
-
-    protected function createEntryPoint($container, $id, $config, $defaultEntryPoint)
```
Comment (dunglas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SimpleFormFactory.php]: Should be reverted, it's a BC break.

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Resources/views/translation.html.php [hunk 3]
```diff
@@ -46,4 +46,4 @@
 
 <?php echo $view['translator']->trans('typecast', ['a' => (int) '123'], 'not_messages'); ?>
 <?php echo $view['translator']->transChoice('msg1', 10 + 1, [], 'not_messages'); ?>
-<?php echo $view['translator']->transChoice('msg2', intval(4.5), [], 'not_messages'); ?>
+<?php echo $view['translator']->transChoice('msg2', (int)4.5, [], 'not_messages'); ?>
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Resources/views/translation.html.php]: The `intval()` call was intended here to have a test case with a function call. We could replace it with another function call, but IMO that's not really necessary as we are only talking about a test fixture here.

File: src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/MongoDbSessionHandlerTest.php [hunk 4]
```diff
@@ -108,7 +108,7 @@ public function testRead()
 
                 if (phpversion('mongodb')) {
                     $that->assertInstanceOf('MongoDB\BSON\UTCDateTime', $criteria[$that->options['expiry_field']]['$gte']);
-                    $that->assertGreaterThanOrEqual(round(intval((string) $criteria[$that->options['expiry_field']]['$gte']) / 1000), $testTimeout);
+                    $that->assertGreaterThanOrEqual(round((int)((string) $criteria[$that->options['expiry_field']]['$gte']) / 1000), $testTimeout);
```
Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/MongoDbSessionHandlerTest.php]: missing space after `(int)`

Comment (xabbuh) [inline src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/MongoDbSessionHandlerTest.php]: same below

File: src/Symfony/Component/HttpKernel/Client.php [hunk 5]
```diff
@@ -48,26 +48,6 @@ public function __construct(HttpKernelInterface $kernel, array $server = array()
     }
 
     /**
-     * {@inheritdoc}
-     *
-     * @return Request|null A Request instance
-     */
-    public function getRequest()
```
Comment (wouterj) [inline src/Symfony/Component/HttpKernel/Client.php]: These 2 methods are intended in order to get correct autocompletion in IDE's. (the base client can only typehint `object`, while the HttpKernel client knows that object is a HttpFoundation request/response).

File: src/Symfony/Component/Security/Acl/Dbal/MutableAclProvider.php [hunk 6]
```diff
@@ -662,9 +662,7 @@ protected function getSelectSecurityIdentityIdSql(SecurityIdentityInterface $sid
     protected function getDeleteSecurityIdentityIdSql(SecurityIdentityInterface $sid)
     {
         $select = $this->getSelectSecurityIdentityIdSql($sid);
-        $delete = preg_replace('/^SELECT id FROM/', 'DELETE FROM', $select);
-
-        return $delete;
+        return preg_replace('/^SELECT id FROM/', 'DELETE FROM', $select);
```
Comment (wouterj) [inline src/Symfony/Component/Security/Acl/Dbal/MutableAclProvider.php]: these should be an empty line before a return statement

**Review by wouterj — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SimpleFormFactory.php (hunk 2 above)
Comment (kalessil) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SimpleFormFactory.php]: Same function with the same body defined in a parent class, so BC is not introduced. 
In nutshell, this is duplicate code if \Symfony\Bundle\SecurityBundle\DependencyInjection\Security\Factory\FormLoginFactory::createEntryPoint

File: src/Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Resources/views/translation.html.php (hunk 3 above)
Comment (kalessil) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Resources/views/translation.html.php]: Hm, I wondering: what is the intention of using the function here, would you share the  idea with me please?

File: src/Symfony/Component/HttpKernel/Client.php (hunk 5 above)
Comment (kalessil) [inline src/Symfony/Component/HttpKernel/Client.php]: I see the point, in this case following approach would work as well:
```php
/**
 * Client simulates a browser and makes requests to a Kernel object.
 *
 * @author Fabien Potencier <fabien@symfony.com>
 *
 * @method Request|null getRequest() A Request instance
 * @method Response|null getResponse() A Response instance
 */
```
Elegant and no additional code needed, what is your opinion aboutit?

File: src/Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Resources/views/translation.html.php (hunk 3 above)
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Resources/views/translation.html.php]: The corresponding test case is processing the PHP token stream. And this one is here to make sure that a function/method call wouldn't break.

Comment (kalessil) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Resources/views/translation.html.php]: Oh, I see. Don't you mind if I'll then use the ceil function instead?

Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Resources/views/translation.html.php]: fine for me

Comment (kalessil) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Fixtures/Resources/views/translation.html.php]: Great, thank you =) Will update today (in the evening most probably)

File: src/Symfony/Component/DomCrawler/Crawler.php [hunk 7]
```diff
@@ -1044,8 +1044,6 @@ private function findNamespacePrefixes($xpath)
      */
     private function createSubCrawler($nodes)
     {
-        $crawler = new static($nodes, $this->uri, $this->baseHref);
-
-        return $crawler;
+        return new static($nodes, $this->uri, $this->baseHref);
```
Comment (stof) [inline src/Symfony/Component/DomCrawler/Crawler.php]: should be reverted to simplify merging between branches (as of 2.8, we have additional code before the return statement)

File: src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/MongoDbSessionHandlerTest.php [hunk 8]
```diff
@@ -167,7 +167,7 @@ public function testWrite()
             $that->assertEquals('bar', $data[$that->options['data_field']]->getData());
             $that->assertInstanceOf('MongoDB\BSON\UTCDateTime', $data[$that->options['time_field']]);
             $that->assertInstanceOf('MongoDB\BSON\UTCDateTime', $data[$that->options['expiry_field']]);
-            $that->assertGreaterThanOrEqual($expectedExpiry, round(intval((string) $data[$that->options['expiry_field']]) / 1000));
+            $that->assertGreaterThanOrEqual($expectedExpiry, round((int) ((string) $data[$that->options['expiry_field']]) / 1000));
```
Comment (stof) [inline src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/MongoDbSessionHandlerTest.php]: does the ``(int)`` casting apply before or after the division by 1000 here ? This code is less readable than before, as it requires knowing the operator precedence (and I cannot guarantee that it works as before here without checking in the php documentation, as I don't know this precedence by heart)

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DomCrawler/Crawler.php (hunk 7 above)
Comment (kalessil) [inline src/Symfony/Component/DomCrawler/Crawler.php]: Will do

File: src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/MongoDbSessionHandlerTest.php (hunk 8 above)
Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/MongoDbSessionHandlerTest.php]: Will add necessary `()` here

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SimpleFormFactory.php (hunk 2 above)
Comment (kalessil) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SimpleFormFactory.php]: ping @dunglas

File: src/Symfony/Component/HttpKernel/Client.php (hunk 5 above)
Comment (kalessil) [inline src/Symfony/Component/HttpKernel/Client.php]: ping @wouterj

Comment (fabpot) [inline src/Symfony/Component/HttpKernel/Client.php]: @kalessil I think adding the annotations is indeed enough. Go for it.

File: src/Symfony/Component/HttpKernel/DataCollector/Util/ValueExporter.php (hunk 1 above)
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/DataCollector/Util/ValueExporter.php]: That code is in a DataCollector, so there is no BC promise here. So, you can update the code.

File: src/Symfony/Component/Yaml/Parser.php [hunk 9]
```diff
@@ -610,7 +610,8 @@ private function parseBlockScalar($style, $chomping = '', $indentation = 0)
             $previousLineIndented = false;
             $previousLineBlank = false;
 
-            for ($i = 0; $i < count($blockLines); ++$i) {
+            $blockLinesCount = count($blockLines);
+            for ($i = 0; $i < $blockLinesCount; ++$i) {
```
Comment (fabpot) [inline src/Symfony/Component/Yaml/Parser.php]: Can you inline the assignment in the `for` code?

File: src/Symfony/Component/HttpFoundation/BinaryFileResponse.php [hunk 10]
```diff
@@ -164,7 +164,8 @@ public function setContentDisposition($disposition, $filename = '', $filenameFal
         if ('' === $filenameFallback && (!preg_match('/^[\x20-\x7e]*$/', $filename) || false !== strpos($filename, '%'))) {
             $encoding = mb_detect_encoding($filename, null, true);
 
-            for ($i = 0; $i < mb_strlen($filename, $encoding); ++$i) {
+            $filenameLength = mb_strlen($filename, $encoding);
+            for ($i = 0; $i < $filenameLength; ++$i) {
```
Comment (fabpot) [inline src/Symfony/Component/HttpFoundation/BinaryFileResponse.php]: Can you inline the assignment in the `for` code?

File: src/Symfony/Component/Yaml/Parser.php (hunk 9 above)
Comment (kalessil) [inline src/Symfony/Component/Yaml/Parser.php]: Yes, will do.

File: src/Symfony/Component/HttpFoundation/BinaryFileResponse.php (hunk 10 above)
Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/BinaryFileResponse.php]: Yes, will do.

**Comment by fabpot** [comment]
👍

**Comment by kalessil** [comment]
Ready now: all green

**Comment by fabpot** [comment]
Thank you @kalessil.

**Pull request closed without merging**